TZone: Implement TZone Allocation on JavaScriptCore Types

msaboff opened the pull request:
#### 40f96223908c9f08584300adf79bc50a855116cc
<pre>
TZone: Implement TZone Allocation on JavaScriptCore Types
<a href="https://bugs.webkit.org/show_bug.cgi?id=268673">https://bugs.webkit.org/show_bug.cgi?id=268673</a>
<a href="https://rdar.apple.com/122213822">rdar://122213822</a>

Reviewed by Keith Miller.

This patch implements TZone selection in bmalloc through the newly added TZoneHeapManager class.
The macro that defines a TZone allocated type puts the type name, type size and type aligment in a static bmalloc_type.
That static object is now placed in a newly added __DATA_CONST$__tzone_descs section.  At startup time, each dylib / framework
will register their types by passing this new section&apos;s start and end location to the TZoneHeapManager::registerTZoneTypes() method.
When allocating a type for the first time, TZoneHeapManager will allocate if needed for all buckets for the same size and alignment.
Then the manager will compute which of those bucket to use the current type.  This is computed by a process start up seed and the
size and type.

The JavaScript types have been annotated and all JavaScript tests run.  DumpRenderTree has been modified as well.

This patch has TZone allocation turned off via BUSE_TZONE in bmalloc/BPlatform.h and USE_TZONE_MALLOC in wtc/PlatformUse.h
The BUSE_TZONE macro was added to work around a linker bug where unused exported functions end up being eliminated.

* Source/JavaScriptCore/API/tests/testapi.c:
(main):
* Source/JavaScriptCore/API/tests/testapi.cpp:
(WTFTZoneInit):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/assembler/testmasm.cpp:
(main):
* Source/JavaScriptCore/b3/air/testair.cpp:
(main):
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(main):
* Source/JavaScriptCore/dfg/testdfg.cpp:
(main):
* Source/JavaScriptCore/jit/JITOpaqueByproducts.h:
* Source/JavaScriptCore/jsc.cpp:
(main):
(jscmain):
* Source/JavaScriptCore/runtime/RegisterTZoneTypes.cpp: Added.
(JSC::registerTZoneTypes):
* Source/JavaScriptCore/runtime/RegisterTZoneTypes.h: Added.
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/PlatformUse.h:
* Source/WTF/wtf/TZoneMalloc.h:
* Source/WTF/wtf/TZoneMallocInitialization.h: Copied from Source/WTF/wtf/TZoneMalloc.h.
* Source/WTF/wtf/TZoneMallocInlines.h:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/BPlatform.h:
* Source/bmalloc/bmalloc/Map.h:
(bmalloc::Map::contains):
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::api::roundUpToMulipleOf8):
(bmalloc::api::TZoneHeap::provideHeap):
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp: Added.
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::initTypenameTemplate):
(bmalloc::api::TZoneHeapManager::registerTZoneTypes):
(bmalloc::api::nameForType):
(bmalloc::api::nameForTypeUpdateIndex):
(bmalloc::api::TZoneHeapManager::closeRegistration):
(bmalloc::api::TZoneHeapManager::ensureInstance):
(bmalloc::api::TZoneHeapManager::populateBucketsForSizeClass):
(bmalloc::api::TZoneHeapManager::heapRefForTZoneType):
* Source/bmalloc/bmalloc/TZoneHeapManager.h: Added.
(bmalloc::api::TZoneHeapManager::SizeAndAlign::SizeAndAlign):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::size const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::alignment const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::key const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::hash):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::operator== const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::operator bool const):
(bmalloc::api::TZoneHeapManager::TZoneHeapManager):
(bmalloc::api::TZoneHeapManager::getInstance):
(bmalloc::api::TZoneHeapManager::mutex):
(bmalloc::api::TZoneHeapManager::bucketCountForTypeCount):
(bmalloc::api::TZoneHeapManager::tzoneBucketForKey):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(DumpRenderTreeMain):

Canonical link: <a href="https://commits.webkit.org/274606@main">https://commits.webkit.org/274606@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/531be2ec74e23690bfad29366baa50e05cceddb7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39525 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42059 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35425 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21386 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15833 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40099 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15592 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34220 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13539 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13514 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35170 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43337 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/32978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35885 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35501 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39300 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39151 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14337 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11816 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37558 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15943 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46157 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8855 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15991 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9414 "Found 3 new JSC binary failures: testair, testapi, testmasm, Found 21593 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_indexOf.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_lastindexof.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort3.js.default, ChakraCore.yaml/ChakraCore/test/Array/concat1.js.default, ChakraCore.yaml/ChakraCore/test/Array/join2.js.default, ChakraCore.yaml/ChakraCore/test/Array/memop_alias.js.default, ChakraCore.yaml/ChakraCore/test/Array/protoLookup.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/shift_unshift.js.default, ChakraCore.yaml/ChakraCore/test/Array/sparsearray.js.default ... (failure)") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15600 "Build is in progress. Recent messages:") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/bmalloc/bmalloc/TZoneHeapManager.cpp:
```diff
@@ -0,0 +1,253 @@
+
+namespace bmalloc { namespace api {
+
+static bool verbose = false;
```
This should probably be `const`

msaboff replied:
I'll make that change.

MenloDorian replied:
`constexpr` would be better.

msaboff replied:
Done.

kmiller68 commented on Source/JavaScriptCore/API/tests/testapi.c:
```diff
@@ -1439,6 +1452,11 @@ int main(int argc, char* argv[])
 
+#if USE(TZONE_MALLOC)
+    UNUSED_PARAM(envp);
+    const char* boothash = _simple_getenv(apple, "executable_boothash");
```
I'm confused why do we get this from `apple` instead of `envp`? Is `apple` some special environment variables? Can we call it `darwinEnvp`, if so?

msaboff replied:
Yes, a separate environment for limited items.  Copied from code that I was given to get the seed.  I can change the name as you suggest.

kmiller68 commented on Source/bmalloc/bmalloc/Map.h:
```diff
@@ -67,6 +65,16 @@ template<typename Key, typename Value, typename Hash> class Map {
+        if (!size())
+            return false;
+
+        auto& bucket = find(key, [&](const Bucket& bucket) { return !bucket.key || bucket.key == key; });
```
Does this work? Your predicate seems wrong. Shouldn't it be `bucket.key == key`? But then `find` looks like it expects the predicate to return true at some point. As it stands now, this function doesn't work if someone ever deletes your key. Which seems to be true for your change but is a bad API in general...

msaboff replied:
This is the same predicate that is in set() on line 60.  It finds either the first bucket that could have the entry we're looking for or the entry.  The return statement on line 75 indicates that the map contains the entry.  This works fine if there isn't any deleting.  Which is the case for the uses in the PR.

It would appear that both set() and contains() have the same flaw if a delete() happens that would put a hole in between where the hash started the search and where it was found with the iterative closed hashing.  I'll fix this for the generic use case.

kmiller68 replied:
I get that but this predicate will return `false` for entries that are still in the map if you happen to delete a key you need to probe over. That is a pretty big footgun.

msaboff replied:
As I said, I'll fix the issue.  The plan is to add another enum param to the template that has `{ DeletingAllowed, DeletingNotAllowed }` and use that to select the predicate and crash is `delete()` is called for the `DeletingNotAllowed` case.

kmiller68 replied:
I don't think `set` has the flaw because holes are ok in `get`.

msaboff replied:
But lets suppose we do `set(x)` and the find loop iterates over a couple of other none hole entries that collide,  We `remove(y)` and it is one of those colliding non hole entries and follow it by a second `set(x)` again.  The second `set(x)` could be put in the hole created by the `remove`.  Now we have two x entries in the `Map`.  If we `rehash`, there is a chance the first x entry replaces the second x entry.

msaboff replied:
My fix for both `set()` and `contains()` is:
`allowDeleting == AllowDeleting::DeletingAllowed ? bucket.key == key : !bucket.key || bucket.key == key;`

I also added a fix for `find()` in the `DeletingAllowed` case to return the first empty bucket after looking at `m_keyCount` of used buckets.

kmiller68 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -0,0 +1,212 @@
+    {
+        ensureInstance();
+        BASSERT(theTZoneHeapManager);
+        return *theTZoneHeapManager;
```
I think you could make this:
```
if (!theTZoneHeapManager)
    ensureInstance();
BASSERT(theTZoneHeapManager);
return *theTZoneHeapManager;
```

msaboff replied:
Agreed.

kmiller68 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -0,0 +1,212 @@
+
+    static const unsigned typeNameLen = 12;
+
+    typedef uint64_t SHA256ResultAsUnsigned[4];
```
Can we make this `typedef uint64_t SHA256ResultAsUnsigned[CC_SHA256_DIGEST_LENGTH / sizeof(uint64_t)]; static_assert(!(CC_SHA256_DIGEST_LENGTH % sizeof(uint64_t)));`

msaboff replied:
Sure.

kmiller68 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -0,0 +1,212 @@
+    {
+        BASSERT(typeCount);
+
+        if (typeCount == 1)
+            return 1;
+        if (typeCount <= 4)
+            return 2;
+        if (typeCount <= 9)
+            return 3;
+        if (typeCount <= 19)
+            return 4;
+
+        return 5;
```
How did we pick these? Might be worth putting in a comment.

msaboff replied:
Initial cut of the bucket count heuristic.  I'll add a comment.

kmiller68 commented on Source/bmalloc/bmalloc/TZoneHeapManager.h:
```diff
@@ -0,0 +1,222 @@
+        // 2-4 => 2
+        // 5-9 => 3
+        // 10-19 => 4
+        // 20+ => 5
```
Nit: you probably don't need this

msaboff replied:
Fixed.

kmiller68 commented on Source/bmalloc/bmalloc/Map.h:
```diff
@@ -94,12 +110,31 @@ template<typename Key, typename Value, typename Hash> class Map {
+                        return bucket;
+                    }
+                }
+            }
```
Nit: As we talked about offline, this could be cleaned up slightly.

msaboff replied:
Fixed.

kmiller68 reviewed (approved):
r=me with nits.

msaboff committed:
TZone: Implement TZone Allocation on JavaScriptCore Types
https://bugs.webkit.org/show_bug.cgi?id=268673
rdar://122213822

Reviewed by Keith Miller.

This patch implements TZone selection in bmalloc through the newly added TZoneHeapManager class.
The macro that defines a TZone allocated type puts the type name, type size and type aligment in a static bmalloc_type.
That static object is now placed in a newly added __DATA_CONST$__tzone_descs section.  At startup time, each dylib / framework
will register their types by passing this new section's start and end location to the TZoneHeapManager::registerTZoneTypes() method.
When allocating a type for the first time, TZoneHeapManager will allocate if needed for all buckets for the same size and alignment.
Then the manager will compute which of those bucket to use the current type.  This is computed by a process start up seed and the
size and type.

The JavaScript types have been annotated and all JavaScript tests run.  DumpRenderTree has been modified as well.

This patch has TZone allocation turned off via BUSE_TZONE in bmalloc/BPlatform.h and USE_TZONE_MALLOC in wtc/PlatformUse.h
The BUSE_TZONE macro was added to work around a linker bug where unused exported functions end up being eliminated.

* Source/JavaScriptCore/API/tests/testapi.c:
(main):
* Source/JavaScriptCore/API/tests/testapi.cpp:
(WTFTZoneInit):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/assembler/testmasm.cpp:
(main):
* Source/JavaScriptCore/b3/air/testair.cpp:
(main):
* Source/JavaScriptCore/b3/testb3.h:
* Source/JavaScriptCore/b3/testb3_1.cpp:
(main):
* Source/JavaScriptCore/dfg/testdfg.cpp:
(main):
* Source/JavaScriptCore/jit/JITOpaqueByproducts.h:
* Source/JavaScriptCore/jsc.cpp:
(main):
(jscmain):
* Source/JavaScriptCore/runtime/RegisterTZoneTypes.cpp: Added.
(JSC::registerTZoneTypes):
* Source/JavaScriptCore/runtime/RegisterTZoneTypes.h: Added.
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/PlatformUse.h:
* Source/WTF/wtf/TZoneMalloc.h:
* Source/WTF/wtf/TZoneMallocInitialization.h: Copied from Source/WTF/wtf/TZoneMalloc.h.
* Source/WTF/wtf/TZoneMallocInlines.h:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/BPlatform.h:
* Source/bmalloc/bmalloc/Map.h:
(bmalloc::Map::contains):
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
* Source/bmalloc/bmalloc/TZoneHeap.h:
(bmalloc::api::roundUpToMulipleOf8):
(bmalloc::api::TZoneHeap::provideHeap):
* Source/bmalloc/bmalloc/TZoneHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeapManager.cpp: Added.
(bmalloc::api::TZoneHeapManager::init):
(bmalloc::api::TZoneHeapManager::initTypenameTemplate):
(bmalloc::api::TZoneHeapManager::registerTZoneTypes):
(bmalloc::api::nameForType):
(bmalloc::api::nameForTypeUpdateIndex):
(bmalloc::api::TZoneHeapManager::closeRegistration):
(bmalloc::api::TZoneHeapManager::ensureInstance):
(bmalloc::api::TZoneHeapManager::populateBucketsForSizeClass):
(bmalloc::api::TZoneHeapManager::heapRefForTZoneType):
* Source/bmalloc/bmalloc/TZoneHeapManager.h: Added.
(bmalloc::api::TZoneHeapManager::SizeAndAlign::SizeAndAlign):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::size const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::alignment const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::key const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::hash):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::operator== const):
(bmalloc::api::TZoneHeapManager::SizeAndAlign::operator bool const):
(bmalloc::api::TZoneHeapManager::TZoneHeapManager):
(bmalloc::api::TZoneHeapManager::getInstance):
(bmalloc::api::TZoneHeapManager::mutex):
(bmalloc::api::TZoneHeapManager::bucketCountForTypeCount):
(bmalloc::api::TZoneHeapManager::tzoneBucketForKey):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(DumpRenderTreeMain):

Canonical link: https://commits.webkit.org/274606@main
